File: go/internal/install/validate.go
package install

import (
	"errors"
	"os"
	"path/filepath"

	"github.com/osolmaz/skillflag/go/internal/core"
)

// SkillMetadata is the minimal SKILL.md frontmatter the installer requires.
type SkillMetadata struct {
	Name        string
	Description string
}

// AssertSkillDir requires SKILL.md at the bundle root.
func AssertSkillDir(rootDir string) error {
	if _, err := os.Stat(filepath.Join(rootDir, "SKILL.md")); err != nil {
		return errors.New("SKILL.md not found in skill root.")
	}
	return nil
}

// ReadSkillMetadata reads and validates SKILL.md frontmatter.
func ReadSkillMetadata(rootDir string) (SkillMetadata, error) {
	content, err := os.ReadFile(filepath.Join(rootDir, "SKILL.md"))
	if err != nil {
		return SkillMetadata{}, err
	}
	fields := core.ParseFrontmatter(string(content))
	name := fields["name"]
	description := fields["description"]

	if name == "" {
		return SkillMetadata{}, errors.New("SKILL.md metadata is missing name.")
	}
	if description == "" {
		return SkillMetadata{}, errors.New("SKILL.md metadata is missing description.")
	}
	return SkillMetadata{Name: name, Description: description}, nil
}

File: rust/crates/skillflag/src/install/validate.rs
use std::fs;
use std::path::Path;

use crate::core::errors::{Result, SkillflagError};
use crate::frontmatter::parse_frontmatter;

#[derive(Debug, Clone)]
pub struct SkillMetadata {
    pub name: String,
    pub description: String,
}

pub fn assert_skill_dir(root_dir: &Path) -> Result<()> {
    if root_dir.join("SKILL.md").exists() {
        Ok(())
    } else {
        Err(SkillflagError::new("SKILL.md not found in skill root."))
    }
}

pub fn read_skill_metadata(root_dir: &Path) -> Result<SkillMetadata> {
    let content = fs::read_to_string(root_dir.join("SKILL.md"))?;
    let fields = parse_frontmatter(&content);
    let name = fields
        .get("name")
        .cloned()
        .ok_or_else(|| SkillflagError::new("SKILL.md metadata is missing name."))?;
    let description = fields
        .get("description")
        .cloned()
        .ok_or_else(|| SkillflagError::new("SKILL.md metadata is missing description."))?;
    Ok(SkillMetadata { name, description })
}
